### ISSUE TITLE
Umat and XFEM

### ISSUE THREAD

**Issue by mikejwg** [TarletonGroup/CrystalPlasticity#44]
Dear Erlap,

We intend to combine CP and XFEM to simulate crack propagation. For simplicity, the MAXPS criteria embedded in abaqus are used. When CPFE and XFEM were tested separately, the program ran normally. But when  combining both, an error occurs. 
abaqus/standard rank 0 encountered a segmentation violation referencing loacation.
 ---------- RUNTIME EXCEPTION HAS OCCURED ----------
ABAQUS/standard rank 0 terminated by signal 11 
*** ERROR CATEGORY: INITIAL STRESS.
Any suggestions are very appreciated.

inp codes: (abaqus 2020)
...
*Elset, elset=_PICKEDSET24, internal, instance=TESS-1, generate
     1,  50653,      1
** Constraint: 
***Note:   Eqn-1  for applying tension displacement 
*Equation
2
SET-2, 1, 1.
SET-1, 1, -1.
*Enrichment, name=CRACK-1, type=PROPAGATION CRACK, elset=_PICKEDSET24, interaction=IntProp-1
*End Assembly
** 
** MATERIALS
** 
**  MATERIALS
*Material, name=MATERIAL-GRAIN1
*Damage Initiation, criterion=MAXPS
1000.,
*Damage Evolution, type=DISPLACEMENT
 1,
*Damage Stabilization
1e-06
*Depvar
    125,
*User Material, constants=6
 59.5614,  125.136, -21.5391,       1.,      11.,       0.
*Material, name=MATERIAL-GRAIN2
*Damage Initiation, criterion=MAXPS
1000.,
*Damage Evolution, type=DISPLACEMENT
 1,
*Damage Stabilization
1e-06
*Depvar
    125,
*User Material, constants=6
 -55.9356, 57.2808, 199.753,      2.,     11.,      0.
...
** INTERACTION PROPERTIES
** 
*Surface Interaction, name=IntProp-1
1.,
*Friction, slip tolerance=0.005
 0.115,
*Surface Behavior, pressure-overclosure=LINEAR
100000., 
*Initial Conditions, type=ENRICHMENT
TESS-1.47933, 1, CRACK-1, -0.540541, 0.0777657
TESS-1.47933, 2, CRACK-1, 1e-06, 0.0777657
TESS-1.47933, 3, CRACK-1, 1e-06, 0.170132
TESS-1.47933, 4, CRACK-1, -0.540541, 0.170132
TESS-1.47933, 5, CRACK-1, -0.540541, -0.44067
TESS-1.47933, 6, CRACK-1, 1e-06, -0.44067
TESS-1.47933, 7, CRACK-1, 1e-06, -0.212088
TESS-1.47933, 8, CRACK-1, -0.540541, -0.212088
TESS-1.47970, 1, CRACK-1, -0.540541, 0.170132
TESS-1.47970, 2, CRACK-1, 1e-06, 0.170132
TESS-1.47970, 3, CRACK-1, 1e-06, 0.552352
TESS-1.47970, 4, CRACK-1, -0.540541, 0.552352
TESS-1.47970, 5, CRACK-1, -0.540541, -0.212088
TESS-1.47970, 6, CRACK-1, 1e-06, -0.212088
TESS-1.47970, 7, CRACK-1, 1e-06, 0.170132
TESS-1.47970, 8, CRACK-1, -0.540541, 0.170132
TESS-1.49339, 1, CRACK-1, -0.540541, -0.212088
TESS-1.49339, 2, CRACK-1, 1e-06, -0.212088
TESS-1.49339, 3, CRACK-1, 1e-06, 0.170132
TESS-1.49339, 4, CRACK-1, -0.540541, 0.170132
TESS-1.49339, 5, CRACK-1, -0.540541, -0.44067
TESS-1.49339, 6, CRACK-1, 1e-06, -0.44067
TESS-1.49339, 7, CRACK-1, 1e-06, 0.0777653
TESS-1.49339, 8, CRACK-1, -0.540541, 0.0777653
TESS-1.49302, 1, CRACK-1, -0.540541
TESS-1.49302, 2, CRACK-1, 1e-06
TESS-1.49302, 3, CRACK-1, 1e-06
TESS-1.49302, 4, CRACK-1, -0.540541
TESS-1.49302, 5, CRACK-1, -0.540541
TESS-1.49302, 6, CRACK-1, 1e-06
TESS-1.49302, 7, CRACK-1, 1e-06
TESS-1.49302, 8, CRACK-1, -0.540541
** ----------------------------------------------------------------
** 
** STEP: Step-1
** 
*Step, name=Step-1, nlgeom=YES, inc=10000
*Static
1., 10., 1e-06, 0.1
** 
** BOUNDARY CONDITIONS
** 
** Name: Disp-BC-1 Type: Displacement/Rotation
*Boundary
"X=0", 1, 1
** Name: Disp-BC-2 Type: Displacement/Rotation
*Boundary
"Y=0", 2, 2
** Name: Disp-BC-3 Type: Displacement/Rotation
*Boundary
"Z=0", 3, 3
** Name: Disp-BC-4 Type: Displacement/Rotation
*Boundary
SET-1, 1, 1, 2.
** 
** INTERACTIONS
** 
** Interaction: Int-1
*Enrichment Activation, name=CRACK-1, activate=ON
** 
** OUTPUT REQUESTS
----

**Comment by mikejwg** [comment]
*Step, name=Step-1, nlgeom=YES, inc=10000
*Static
0.1, 10., 1e-06, 0.1

**Comment by EralpDemir** [comment]
Hi Mike,
Nice work.

Please check if the use of XFEM changes  the number of elements during the analysis or not. If it does so, you may try to set a greater number for the parameter "numel" in userinputs.

The UMAT must use a fixed number of element for GND calculation. So, GND analysis shall be turned off in this case.

Let me know how it works,
Eralp

**Comment by mikejwg** [comment]
Dear Eralp,
Thanks for your suggestion.
We turned off the GND analysis (meanning without grain size effect ?) and set a greater number of numel in userinputs (about 10 times the total number of elements），although we thought in the crack propagation process of XFEM, the number of elements will not increase, only the nodes on the propagation path are split into two nodes.  Our tests shown the same error message. 

GND calculation identifier
!     0: No GNDs
!     1: (curlxFp)^T followed L2 method - SVD - Active SS restriction (new)
!     2: Gurtin's measure for incompatibility - L2 - SVD - Active SS restriction (Cermeli, P. and Gurtin, M.E., 2001)
!     3: curl(gdotnFp) followed by projections (H.Dai, PhD Thesis, 1997.)
!     4: slip gradients followed by projections (Gerken, J.M. and Dawson, P.R., 2008.)
!     Default value is set to 0
      integer, parameter, public :: gndmodel = 0

Thanks.

Mike

**Comment by mikejwg** [comment]
if (LOP==0) then
!
          write(*,*) '--------------------------------'
!
!         Call the initializations
          call initialize_once
!
!         Display upon completion
          write(*,*) 'initialization once at the
     + beginning is complete!'
!
!
          write(*,*) '--------------------------------'
!
      endif

_The code exits here after executing the above command.
The following command is not executed._


!     At the end of each increment
      if (LOP==2) then

**Comment by mikejwg** [comment]
we debug UMAT.f by writing some varibles  (v2.5 was downloaded). 
write(*,*) '1--------------------------------',noel,npt.
If 'CPE4' was used in the simulation (, we found that,  during the running, the npt may change to 5. 
ABAQUS element type: CPE4
 Linear interpolation for will be used GNDs
 Dimension of the analysis:            2
 Number of integration points per element:            4
 Number of nodes per element:            4
 1. Mesh initialization completed!
 2. Array allocation completed!
 3. Identity tensors initialized!
 4. Outputs are defined using useroutputs.f!
 initialization once at the beginning is complete!
 --------------------------------
 1--------------------------------           1           1
1 umat--------------------------------           1           4
 1 umat 1--------------------------------
 1 umat 2--------------------------------
 1--------------------------------           1           5
 
 1 umat--------------------------------           1           5
 1 umat 1--------------------------------
 1 umat 2--------------------------------
 1--------------------------------           1           6
 2--------------------------------
 3--------------------------------
 4--------------------------------
 5--------------------------------
 6--------------------------------
 7--------------------------------
 8--------------------------------
 9--------------------------------
 12--------------------------------
 13--------------------------------
 14--------------------------------           1           6
 15--------------------------------
 151--------------------------------
 152--------------------------------
 153--------------------------------           1           6
 154--------------------------------           1           6
 155--------------------------------           1           6
 156--------------------------------           1           6
 157--------------------------------           1           6
 158--------------------------------
 159--------------------------------           1           6
 1515--------------------------------           1           6        
***  ABAQUS/standard rank 0 encountered a SEGMENTATION FAULT

*** ERROR CATEGORY:  INITIAL STRESS

**Comment by mikejwg** [comment]
Dear Eralp,
I got it.  Because the number of integration points  used in XFEM is greater than that not used in  XFEM, I changed values in the original code. The code works fine now.
Thanks Eralp.

Best, 
Mike

**Comment by Wang-Yang-51** [comment]
Dear guys，
I am using a UMAT and UDMGINI subroutine in Abaqus, combined with XFEM, to simulate fatigue crack propagation. The model runs without errors, but the crack does not propagate in the results.  
I have tried adjusting analysis steps, boundary conditions, grain crystal orientations, and the damage initiation criterion in the UDMGINI subroutine, but the results are still unsatisfactory. I am unsure how to further modify my model and would appreciate any guidance or assistance.

![Image](https://github.com/user-attachments/assets/b26f5033-0309-4d61-bbab-be9ea630de40)
![Image](https://github.com/user-attachments/assets/77bdcde8-b819-4496-85d3-12022e8f92a5)
![Image](https://github.com/user-attachments/assets/7d4e7d29-82ea-4599-8e1c-97a8350949c1)

**Comment by EralpDemir** [comment]
Hi, 
We will have a version coupled with UDMGINI subroutine. It will be available after publication.

Best wishes,
Eralp

**Comment by Wang-Yang-51** [comment]
> Hi, We will have a version coupled with UDMGINI subroutine. It will be available after publication.
> 
> Best wishes, Eralp

That’s great! Thank you so much.

**Comment by Wang-Yang-51** [comment]
Dear all,  
I've encountered a new issue and would like to ask: how can I extract information about the crack growth rate and the number of cycles from the results? In my model, the crack propagates to the end in just 27 step times, which I believe is too few and not realistic. However, it's also impractical to simulate tens of thousands of cycles. I would really appreciate any help or suggestions you might have./

![Image](https://github.com/user-attachments/assets/14fb744e-342c-4e09-83e4-c6d69fc84fed)

**Comment by EralpDemir** [comment]
Thanks for sharing your results. You may check out the Abaqus documentation for that.
Best
